crates/bevy_math/src/shape_sampling.rs

```diff
@@ -0,0 +1,239 @@
+    ///
+    /// # Example
+    /// ```
+    /// # use bevy_math::prelude::*;
+    /// let square = Rectangle::new(2.0, 2.0);
+    ///
+    /// // Returns a Vec2 with both x and y between -1 and 1.
+    /// println!("{:?}", square.sample_volume(&mut rand::thread_rng()));
+    /// ```
+    fn sample_volume<R: Rng + ?Sized>(&self, rng: &mut R) -> Self::Output;
+
+    /// Uniformly sample a point from the surface of this shape, centered on 0.
+    ///
+    /// # Example
+    /// ```
+    /// # use bevy_math::prelude::*;
+    /// let square = Rectangle::new(2.0, 2.0);
+    ///
+    /// // Returns a Vec2 where one of the coordinates is at ±1,
+    /// //  and the other is somewhere between -1 and 1.
+    /// println!("{:?}", square.sample_surface(&mut rand::thread_rng()));
+    /// ```
+    fn sample_surface<R: Rng + ?Sized>(&self, rng: &mut R) -> Self::Output;
+}
+
+impl ShapeSample for Circle {
+    type Output = Vec2;
+
+    fn sample_volume<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec2 {
+        // https://mathworld.wolfram.com/DiskPointPicking.html
+        let theta = rng.gen_range(0.0..TAU);
+        let r_squared = rng.gen_range(0.0..=(self.radius * self.radius));
+        let r = r_squared.sqrt();
+        Vec2::new(r * theta.cos(), r * theta.sin())
+    }
+
+    fn sample_surface<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec2 {
+        let theta = rng.gen_range(0.0..TAU);
+        Vec2::new(self.radius * theta.cos(), self.radius * theta.sin())
+    }
+}
+
+impl ShapeSample for Sphere {
+    type Output = Vec3;
+
+    fn sample_volume<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec3 {
+        // https://mathworld.wolfram.com/SpherePointPicking.html
+        let theta = rng.gen_range(0.0..TAU);
+        let phi = rng.gen_range(-1.0_f32..1.0).acos();
+        let r_cubed = rng.gen_range(0.0..=(self.radius * self.radius * self.radius));
+        let r = r_cubed.cbrt();
+        Vec3 {
+            x: r * phi.sin() * theta.cos(),
+            y: r * phi.sin() * theta.sin(),
+            z: r * phi.cos(),
+        }
+    }
+
+    fn sample_surface<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec3 {
+        let theta = rng.gen_range(0.0..TAU);
+        let phi = rng.gen_range(-1.0_f32..1.0).acos();
+        Vec3 {
+            x: self.radius * phi.sin() * theta.cos(),
+            y: self.radius * phi.sin() * theta.sin(),
+            z: self.radius * phi.cos(),
+        }
+    }
+}
+
+impl ShapeSample for Rectangle {
+    type Output = Vec2;
+
+    fn sample_volume<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec2 {
+        let x = rng.gen_range(-self.half_size.x..=self.half_size.x);
+        let y = rng.gen_range(-self.half_size.y..=self.half_size.y);
+        Vec2::new(x, y)
+    }
+
+    fn sample_surface<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec2 {
+        let side_distance = rng.gen_range(-1.0..1.0);
+        let other_side = rng.gen_range(0..=1) as f32 * 2.0 - 1.0;
```

Comment:
Here we are generating an integer range to then cast into `f32`, with a range of `-1.0..=1.0`. Might be better to just do `rng.gen_range(-1.0f32..=1.0)`?

Reply:
We're getting the integer range so that we either get `0` or `1` which then gets cast so that it is either `-1` or `1`. It might be clearer just to do `let other_side = if rng.gen_ratio(1, 2) { -1.0 } else { 1.0 }` though I'm not sure

Reply:
Maybe just generate a bool? You could look at compiler output, it's probably a similar number of operations.

Reply:
For what it's worth, I also found this confusing to read without type annotations.

Reply:
Agreed.

Reply:
Yeah, it wasn't particularly nice, I've changed both instances of this to `rng.gen_ratio(1, 2)` one-line if statements, although I'm not sure if `rng.gen_ratio(1, 2)` or `rng.gen_bool(0.5)` is cleaner

Reply:
I think you can just use .gen()

Reply:
I had forgotten about that method, yes, thank you, I've changed all `gen_ratio(1, 2)` over to `gen()`